src/commands/synthetics/reporters/junit.ts

```diff
@@ -0,0 +1,280 @@
+      this.destination += '.xml'
+    }
+    this.builder = new Builder()
+    this.json = {
+      testsuites: {$: {name: command.runName || 'Undefined run'}, testsuite: []},
+    }
+  }
+
+  public async runEnd() {
+    try {
+      await fs.writeFile(this.destination.replace(/\.xml$/, '.json'), JSON.stringify(this.json, undefined, 4), 'utf8')
+      const xml = this.builder.buildObject(this.json)
+      await fs.writeFile(this.destination, xml, 'utf8')
+      this.write(`✅ Created a jUnit report at ${c.bold.green(this.destination)}\n`)
+    } catch (e) {
+      this.write(`❌ Couldn't write the report to ${c.bold.green(this.destination)}:\n${e.toString()}\n`)
+    }
+  }
+
+  public testEnd(test: Test, results: PollResult[]) {
+    const suiteRunName = test.suite || 'Undefined suite'
+
+    let suiteRun = this.json.testsuites.testsuite.find((suite: any) => suite.$.name === suiteRunName)
+    if (!suiteRun) {
+      suiteRun = {
+        $: {name: suiteRunName, ...getDefaultStats()},
+        testsuite: [],
+      }
+      this.json.testsuites.testsuite.push(suiteRun as XMLRun)
+    }
+
+    // Update stats for the suite.
+    suiteRun.$ = {
+      ...suiteRun.$,
+      ...this.getSuiteStats(results, getStats(suiteRun.$)),
+    }
+
+    for (const result of results) {
+      const testSuite: XMLSuite = this.getTestSuite(test, result)
+
+      for (const stepDetail of result.result.stepDetails) {
+        testSuite.testcase.push(...this.getStep(stepDetail))
+      }
+
+      suiteRun.testsuite.push(testSuite)
+    }
+  }
+
+  private getResultStats(result: PollResult, stats: Stats | undefined = getDefaultStats()): Stats {
+    const steps = result.result.stepDetails
+      .map((step) => {
+        if (!step.subTestStepDetails) {
+          return [step]
+        }
+
+        return [step, ...step.subTestStepDetails]
+      })
+      .reduce((acc, val) => acc.concat(val), [])
+
+    for (const step of steps) {
+      const stepStats = this.getStepStats(step)
+
+      stats.tests += stepStats.tests
+      stats.errors += stepStats.errors
+      stats.failures += stepStats.failures
+      stats.skipped += stepStats.skipped
+      stats.allowfailures += stepStats.allowfailures
+      stats.assertions += stepStats.assertions
+      stats.warnings += stepStats.warnings
+    }
+
+    return stats
+  }
+
+  private getStep(stepDetail: Step): XMLStep[] {
+    const mainStep: XMLStep = {
+      $: {
+        allow_failure: stepDetail.allowFailure,
+        is_skipped: stepDetail.skipped,
+        name: stepDetail.description,
+        substep_public_id: stepDetail.subTestPublicId,
+        time: stepDetail.duration / 1000,
+        type: stepDetail.type,
+        url: stepDetail.url,
+        ...this.getStepStats(stepDetail),
+      },
+      browser_error: [],
+      error: [],
+      warning: [],
+    }
+    const steps = [mainStep]
+
+    if (stepDetail.subTestStepDetails && stepDetail.subTestStepDetails.length) {
```

Comment:
do you need the `.length` check?

Reply:
Good catch, probably not since we're directly looping.